### DISCUSSION TITLE
Upgrade to C++20

### DISCUSSION THREAD

**Discussion by alecjacobson** [libigl/libigl#2379] [Ideas]
libigl assumes C++11 (published 13 years ago).

C++20 (published 4 years ago) has a bunch of nice features that make coding in C++ easier and more fun. It seems [well supported](https://en.cppreference.com/w/cpp/compiler_support#cpp20). As the maintainer of libigl, I'd prefer that libigl was easier to maintain.

A couple questions for users of libigl:

 - would you stop using libigl if it depended on C++20?
 - would depending on C++14 or C++17 instead change that decision?
 - would you be more likely to help me maintain libigl if it depended on C++20? (🌈 wishful thinking I know!)

**Comment by jdumas** [comment]
Afaik at Adobe very few of our products are using C++20, but C++17 is widely used across the board. What feature from C++20 are you interested in specifically? From what I can read in this table support for modules seems to be lacking, and concepts are not fully supported by AppleClang yet.

**Reply by alecjacobson** [reply]
Thanks! This is really helpful to know. Things I know I love:

- **[`if constexpr` (C++17)](https://en.cppreference.com/w/cpp/language/if) (C++17)** this one in particular would make my libigl-life easier
- [`if constexpr` (C++17)](https://en.cppreference.com/w/cpp/language/if) (C++17)
- [File system library](https://en.cppreference.com/w/cpp/filesystem) (C++17)
- [Structured Bindings](https://en.cppreference.com/w/cpp/language/structured_binding) (C++17)
- [Fold Expressions](https://en.cppreference.com/w/cpp/language/fold) (C++17)
- [template-parameter-list for generic lambdas](https://en.cppreference.com/w/cpp/language/lambda#Syntax) (C++20)
- [Generic lambda expressions](https://en.cppreference.com/w/cpp/language/lambda#Explanation) (C++14)
- [std::is_same_v](https://en.cppreference.com/w/cpp/types/is_same) (C++17)
- [std::apply](https://en.cppreference.com/w/cpp/utility/apply) (C++17)

**Reply by jdumas** [reply]
Based on this list, it seems like moving to C++17 will be a safer bet and covers most of your needs. template-parameter list for generic lambdas is nice-to-have but it's not something I find the need to do very often I think?

**Comment by Shuenhoy** [comment]
Another useful feature brought by C++17 is `static inline`. It can reduce some unnecessarily repetitive code:
```diff
diff --git a/include/igl/WindingNumberTree.h b/include/igl/WindingNumberTree.h
index 91d084e6c..eb83f10c1 100644
--- a/include/igl/WindingNumberTree.h
+++ b/include/igl/WindingNumberTree.h
@@ -27,13 +27,13 @@ namespace igl
     public:
       // Method to use (see enum above)
       //static double min_max_w;
-      static std::map< 
+      static inline std::map< 
         std::pair<const WindingNumberTree*,const WindingNumberTree*>, 
         typename DerivedV::Scalar>
           cached;
       // This is only need to fill in references, it should never actually be touched
       // and shouldn't cause race conditions. (This is a hack, but I think it's "safe")
-      static DerivedV dummyV;
+      static inline DerivedV dummyV;
     protected:
       WindingNumberMethod method;
       const WindingNumberTree * parent;
@@ -153,9 +153,6 @@ namespace igl
 //WindingNumberMethod WindingNumberTree<Point,DerivedV,DerivedF>::method = EXACT_WINDING_NUMBER_METHOD;
 //template <typename Point, typename DerivedV, typename DerivedF>
 //double WindingNumberTree<Point,DerivedV,DerivedF>::min_max_w = 0;
-template <typename Point, typename DerivedV, typename DerivedF>
-std::map< std::pair<const igl::WindingNumberTree<Point,DerivedV,DerivedF>*,const igl::WindingNumberTree<Point,DerivedV,DerivedF>*>, typename DerivedV::Scalar>
-  igl::WindingNumberTree<Point,DerivedV,DerivedF>::cached;
  
 template <typename Point, typename DerivedV, typename DerivedF>
 inline igl::WindingNumberTree<Point,DerivedV,DerivedF>::WindingNumberTree():
@@ -491,11 +488,4 @@ igl::WindingNumberTree<Point,DerivedV,DerivedF>::cached_winding_number(
   return 0;
 }
  
-// Explicit instantiation of static variable
-template <
-  typename Point,
-  typename DerivedV, 
-  typename DerivedF >
-DerivedV igl::WindingNumberTree<Point,DerivedV,DerivedF>::dummyV;
-
 #endif
```

This will be also useful if there will a wrapper for c++20 modules in the future. Only applying the above patch can this module wrapper compile.
```c++
module;
#include <igl/winding_number.h>
#include <igl/signed_angle.h>
#include <igl/readOBJ.h>
#include <igl/writeOBJ.h>
// ...

export module igl.core;

export namespace igl {

using igl::signed_angle;
using igl::winding_number;
using igl::writeOBJ;
using igl::readOBJ;
// ...
} // namespace igl
```

**Comment by fire** [comment]
Please do not update past C++ 2017. I have dependent projects like Godot Engine locked at that C++ version.
